unified-runtime/source/adapters/level_zero/v2/queue_batched.cpp

```diff
@@ -0,0 +1,974 @@
+
+#include "../common/latency_tracker.hpp"
+#include "../helpers/kernel_helpers.hpp"
+#include "../image_common.hpp"
+
+#include "../program.hpp"
+#include "../ur_interface_loader.hpp"
+#include "ur_api.h"
+#include "ze_api.h"
+#include <cstddef>
+#include <cstdint>
+#include <tuple>
+
+namespace v2 {
+
+ur_queue_batched_t::ur_queue_batched_t(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice, uint32_t ordinal,
+    ze_command_queue_priority_t priority, std::optional<int32_t> index,
+    [[maybe_unused]] event_flags_t eventFlags, ur_queue_flags_t flags)
+    : regularCmdListDesc(v2::command_list_desc_t{
+          true /* isInOrder*/, ordinal /* Ordinal*/,
+          true /* copyOffloadEnable*/, false /*isMutable*/}),
+      currentCmdLists(
+          hContext, hDevice,
+          /* regular command list*/
+          hContext->getCommandListCache().getRegularCommandList(
+              hDevice->ZeDevice, regularCmdListDesc),
+          /* command list immediate*/
+          hContext->getCommandListCache().getImmediateCommandList(
+              hDevice->ZeDevice,
+              {true, ordinal, true /* always enable copy offload */},
+              ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS, priority, index)
+
+      ) {
+  TRACK_SCOPE_LATENCY("ur_queue_batched_t::constructor");
+
+  // TODO common code?
+  if (!hContext->getPlatform()->ZeCommandListImmediateAppendExt.Supported) {
+    UR_LOG(ERR, "Adapter v2 is used but the current driver does not support "
+                "the zeCommandListImmediateAppendCommandListsExp entrypoint.");
+    throw UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  }
+
+  this->hContext = hContext;
+  this->hDevice = hDevice;
+
+  this->flags = flags;
+
+  eventPoolRegular = hContext->getEventPoolCache(PoolCacheType::Regular)
+                         .borrow(hDevice->Id.value(), v2::EVENT_FLAGS_COUNTER);
+}
+
+ur_event_handle_t ur_queue_batched_t::createEventIfRequestedRegular(
+    ur_event_handle_t *phEvent, ur_event_generation_t batch_generation) {
+  TRACK_SCOPE_LATENCY("ur_queue_batched_t::createEventIfRequested");
+
+  if (phEvent == nullptr) {
+    return nullptr;
+  }
+
+  (*phEvent) = eventPoolRegular->allocate();
+  (*phEvent)->setQueue(this);
+  (*phEvent)->setBatch(batch_generation);
+
+  return (*phEvent);
+}
+
+ur_event_handle_t ur_queue_batched_t::createEventAndRetainRegular(
+    ur_event_handle_t *phEvent, ur_event_generation_t batch_generation) {
+  auto hEvent = eventPoolRegular->allocate();
+  hEvent->setQueue(this);
+  hEvent->setBatch(batch_generation);
+
+  if (phEvent) {
+    (*phEvent) = hEvent;
+    hEvent->retain();
+  }
+
+  return hEvent;
+}
+
+ur_result_t batch_manager::renewRegularUnlocked(
+    v2::raii::command_list_unique_handle &&newRegularBatch) {
+  TRACK_SCOPE_LATENCY("batch_manager::renewRegularUnlocked");
+
+  regularGenerationNumber++;
+
+  // save the previous regular for execution
+  // renew regular
+  runBatches.push_back(activeBatch.releaseCommandList());
```

Comment:
You never do anything with these batches that I can find. The simplest option would be to release all the runBatches back to the cache after queue finish (because we are guaranteed they are finished after synchronize).